### PR TITLE
MoveInfo rearrangement and flag optimisation

### PR DESCRIPTION
**Pull request by cfmnephrite**
## Description
Basically, a reordering of the BattleMove struct into a more sensible, more efficient structure.
- Grouped BP/accuracy and type/category together (finally)
- Removed unnecessary flags
    - Made some into EFFECT flags
    - Rewrote code to take modified flags into account
    - Notably, removed Sheer Force flag and used additionalEffects field to determine if a move should be boosted or not
- removed numAdditionalEffects and used some bit trickery to fit a counter for up to 15 additional effects inside the `additionalEffects` field
- Rearranged all moves_info.h data to be to be in struct order

## **Discord contact info**
The Cancer Fairy/thechurchofcage

### CODE REVIEW

File: test/battle/move_effect/upper_hand.c
Comment (AlexOn1ine) [inline test/battle/move_effect/upper_hand.c:102]: It's because it can't be added together with all the other sheer force tests.

Should have asked in the dev channel. Now you have to waste time removing the comment. 😅

Comment (AlexOn1ine) [reply test/battle/move_effect/upper_hand.c:102]: Also conficts

**Comment by cfmnephrite** [comment]
Conflicts resolved and `moves_info.h` information all sorted in struct order (since CONFIG_BY_GEN isn't likely to be resolved by release day). Ready for review :)

**Comment by AlexOn1ine** [comment]
@AsparagusEduardo I think this should make it into 1.8, not 1.9 since this is the last change for the move refactor.

**Comment by cfmnephrite** [comment]
Forgive the very late update, had another look and all of a sudden, BP/accuracy/type/category looked wrong to me compared to type/category/BP/accuracy :P It also matches their order on places like Showdown and Bulbapedia (except PP, but oh, well).

**Comment by AsparagusEduardo** [comment]
> @AsparagusEduardo I think this should make it into 1.8, not 1.9 since this is the last change for the move refactor.

Alright, I'll take a look before release.

**Comment by AlexOn1ine** [comment]
> > @AsparagusEduardo I think this should make it into 1.8, not 1.9 since this is the last change for the move refactor.
> 
> Alright, I'll take a look before release.

If you want I can review it. I was holding back because I thought you might want to take a look.

File: test/battle/ability/sheer_force.c
Comment (AsparagusEduardo) [inline test/battle/ability/sheer_force.c:9]: Why an assumption specifically for Electro Shot?

File: src/data/moves_info.h
Comment (AsparagusEduardo) [inline src/data/moves_info.h:132]: Ok, I need to ask this.

*Why?*

Why change the stablished order? I get grouping banned flags together, but this will create so many conflicts across the board that I don't get why you chose to do this.

File: include/pokemon.h
Comment (AsparagusEduardo) [inline include/pokemon.h:456]: Is there a move with more than 255 power?

File: include/pokemon.h
Comment (AsparagusEduardo) [inline include/pokemon.h:470]: Add this padding as a field.

File: src/battle_ai_main.c
Comment (AsparagusEduardo) [inline src/battle_ai_main.c:2510]: Shouldn't this still check if it's a 2-turn move?

File: include/battle.h
Comment (AsparagusEduardo) [inline include/battle.h:60]: I feel `forcePressure` should be set for moves instead of effects.

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: src/battle_ai_main.c
Comment (AlexOn1ine) [reply src/battle_ai_main.c:2510]: I think the idea is for the AI to charge and fire the move in the same turn if instructed so it would make sense to remove it.

Comment (ZnogyroP) [reply src/battle_ai_main.c:2510]: Two-turn moves are fully banned from Instruct, even if it could be executed it one turn.

Comment (AlexOn1ine) [reply src/battle_ai_main.c:2510]: oh then it even makes more sense to remove it.

Comment (AsparagusEduardo) [reply src/battle_ai_main.c:2510]: > oh then it even makes more sense to remove it.

What? It makes more sense to *not* remove it in this case then, replaced by the effect twoTurnMove flag

Comment (AlexOn1ine) [reply src/battle_ai_main.c:2510]: I mean hopefully those moves have `instructBanned` on them already

File: include/pokemon.h
Comment (cfmnephrite) [reply include/pokemon.h:470]: It's not padding though. Like it says, it just continues into flags. It's just a note for anyone who would like to add an extra field in that position.

File: include/pokemon.h
Comment (cfmnephrite) [reply include/pokemon.h:456]: Not in vanilla, but I don't see why we shouldn't allow for that possibility in romhacks. Besides, it's not like we need up to 255 accuracy.

File: src/data/moves_info.h
Comment (cfmnephrite) [reply src/data/moves_info.h:132]: To put everything into a more sensible order and make better use of space. You approved this when I asked about it ages ago?

File: test/battle/ability/sheer_force.c
Comment (cfmnephrite) [reply test/battle/ability/sheer_force.c:9]: Yup, Electro Shot uses a "hack" to be Sheer Force boosted so it's just a check for that.

File: src/data/moves_info.h
Comment (cfmnephrite) [reply src/data/moves_info.h:132]: Screenshot_20240213-081302_Discord (1)

Tbf, I think you were more interested in the flag rearrangement, but all of the above was in the proposal, for which you said this ^

Comment (cfmnephrite) [reply src/data/moves_info.h:132]: Besides, we've cause conflicts before by renaming `split` (and then also the defines - twice), adding contest data and names, and adding `additionalEffects`. This would probably add more conflicts than any of those but it would be worth it to establish, IMO, now that all move data fields have been added to the struct, a final, clean, sensible order.

Comment (AsparagusEduardo) [reply src/data/moves_info.h:132]: > Tbf, I think you were more interested in the flag rearrangement, but all of the above was in the proposal, for which you said this ^

From what I remember, you only showed the struct rearrangement, not that you were going to change the field order in the data as well.

Comment (AsparagusEduardo) [reply src/data/moves_info.h:132]: > Besides, we've cause conflicts before by renaming `split` (and then also the defines - twice), adding contest data and names, and adding `additionalEffects`. This would probably add more conflicts than any of those but it would be worth it to establish, IMO, now that all move data fields have been added to the struct, a final, clean, sensible order.

The current order is fine. Vanilla fields on top, new fields at the bottom.

File: include/pokemon.h
Comment (AsparagusEduardo) [reply include/pokemon.h:456]: Then we don't need to move the extra bit.

File: src/data/moves_info.h
Comment (cfmnephrite) [reply src/data/moves_info.h:132]: Oh! I thought that having moves info and the struct be in synch was preferred. Has that not been the case?

File: include/battle.h
Comment (AsparagusEduardo) [reply include/battle.h:60]: Which means that it could be interpreted either way, and when that happens, I rather keep it for moves to allow more customization.

File: src/data/moves_info.h
Comment (cfmnephrite) [reply src/data/moves_info.h:132]: In fact, I seem to recall with the species form refactor how you didn't like that the order of the fields had been lost (and you were satisfied when I then send that I would rearrange them again before finishing).

File: include/pokemon.h
Comment (AsparagusEduardo) [reply include/pokemon.h:470]: You could add a space to delimit where the word end instead (or a comment)

File: src/data/moves_info.h
Comment (cfmnephrite) [reply src/data/moves_info.h:132]: > The current order is fine to me. Vanilla fields on top, new fields at the bottom.

They aren't really, tbh. The zMove field is way up there, and the flags are in kind of a random order as well (not that I'm willing to touch those but still).

File: include/pokemon.h
Comment (cfmnephrite) [reply include/pokemon.h:456]: I can't without reordering things again

File: src/data/moves_info.h
Comment (AsparagusEduardo) [reply src/data/moves_info.h:132]: > In fact, I seem to recall with the species form refactor how you didn't like that the order of the fields had been lost (and you were satisfied when I then said that I would rearrange them again before finishing).

I don't recall this much. In the end, the struct was arranged to minimize space used but then the fields were organized such that Dex data and graphics data were all next to each other, despite not that being how it's setup in the struct.

Comment (cfmnephrite) [reply src/data/moves_info.h:132]: @AsparagusEduardo This is what you said back then and it's why I spent my time reordering moves_info to match the struct:
Screenshot_20240213-090257_Edge (1)

Comment (AsparagusEduardo) [reply src/data/moves_info.h:132]: Because this was an automated thing you did that moved fields that were previously in species macros to the top of the array. All mon that didn't have macros kept their previous order, so it was inconsistent.

Comment (AsparagusEduardo) [reply src/data/moves_info.h:132]: In fact, the macros I made were curated to keep the field order that I had established for all non-macro Pokémon, so base stats on top and so on.

Comment (AsparagusEduardo) [reply src/data/moves_info.h:132]: I need to gather my thoughts. I've been outside of the house the whole day and I can't focus.

**Comment by AlexOn1ine** [comment]
CF, your changes look great overall but since Edu isn't sure on the reorder in `moves_info.h` could you revert that part? It would be nice to get this merged before 1.8.

**Comment by AlexOn1ine** [comment]
> CF, your changes look great overall but since Edu isn't sure on the reorder in `moves_info.h` could you revert that part? It would be nice to get this merged before 1.8.

+ conflicts :eyes:

File: src/data/moves_info.h
Comment (cfmnephrite) [reply src/data/moves_info.h]: Reverted the change since we can do it later anyway

**Comment by cfmnephrite** [comment]
@AlexOn1ine Done, just waiting for CI

**Comment by AlexOn1ine** [comment]
Rename suggestions. 
```
RETURN_MOVE_HAS_MOVE_EFFECT_WITH -> RETURN_HAS_MOVE_ADDITIONAL_EFFECT
MoveHasMoveEffect -> MoveHasAdditionalEffect
MoveHasMoveEffectWithChance -> MoveHasAdditionalEffectWithChance 
MoveHasMoveEffectSelf -> MoveHasAdditionalEffectSelf
MoveHasMoveEffectSelfArg -> MoveHasAdditionalEffectSelfArg
MoveHasChargeTurnMoveEffect -> MoveHasChargeTurnAdditionalEffect
```

**Comment by cfmnephrite** [comment]
> MoveHasMoveEffect

Done

**Review by AlexOn1ine — APPROVED** [review]
@AsparagusEduardo From my side it can be merged. The review was mostly done on discord.

File: test/battle/move.c
Comment (AsparagusEduardo) [inline test/battle/move.c:30]: Hmm, shouldn't this test technically still pass if the secondary effect chances are changed by the user? From the name of the test, I assume we want to make sure that the chance modifier does change the chances, so I don't think hardcoding the chances in the assume here is a good idea.
```suggestion
    ASSUME(MoveHasAdditionalEffect(move, MOVE_EFFECT_PARALYSIS) == TRUE);
```

File: include/pokemon.h
Comment (AsparagusEduardo) [inline include/pokemon.h:534]: I recommend moving `GET_ADDITIONAL_EFFECTS_AND_COUNT` to the bottom.
```suggestion
#define GET_ADDITIONAL_EFFECTS(move) (void *)(gMovesInfo[move].additionalEffects & 0x8FFFFFF)
#define GET_ADDITIONAL_EFFECTS_COUNT(move) (gMovesInfo[move].additionalEffects >> 28)
#define GET_ADDITIONAL_EFFECTS_AND_COUNT(move, _count, _effects) u32 _count = GET_ADDITIONAL_EFFECTS_COUNT(move); const struct AdditionalEffect *_effects = GET_ADDITIONAL_EFFECTS(move)
```

File: src/battle_util.c
Comment (AsparagusEduardo) [inline src/battle_util.c:11087]:
```suggestion
#define RETURN_HAS_MOVE_ADDITIONAL_EFFECT(condition1, ...)          \
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]
Besides some small adjustements, looks good to me :)

File: include/pokemon.h
Comment (AsparagusEduardo) [reply include/pokemon.h:534]: oh, and rename `GET_ADDITIONAL_EFFECTS_COUNT` to `GET_ADDITIONAL_EFFECT_COUNT`

File: src/battle_util.c
Comment (cfmnephrite) [reply src/battle_util.c:11087]: What am I supposed to be seeing here? 🤔

Comment (cfmnephrite) [reply src/battle_util.c:11087]: What  I supposed to be changing here?
Screenshot_20240226-205859_Edge (1)

Comment (AsparagusEduardo) [reply src/battle_util.c:11087]: Alignment
image

**Comment by cfmnephrite** [comment]
Suggestions done

**Comment by AsparagusEduardo** [comment]
PR is not passing on the latest upcoming

**Comment by cfmnephrite** [comment]
It looks like TIMEOUT isn't caught by KNOWN_FAILING, for whatever reason. Just fixed the test and removed KNOWN_FAILING.

**Comment by mrgriffin** [comment]
> It looks like TIMEOUT isn't caught by KNOWN_FAILING, for whatever reason. Just fixed the test and removed KNOWN_FAILING.

Glad it's fixed! The reason why `TIMEOUT` isn't caught is because t's not known whether it's a pass or failure, because we don't know that if we ran the test for long enough it wouldn't eventually pass (i.e. it's a variant of the halting problem). Or from an alternative POV, it's not a failure: it's an error.

**Comment by cfmnephrite** [comment]
> > It looks like TIMEOUT isn't caught by KNOWN_FAILING, for whatever reason. Just fixed the test and removed KNOWN_FAILING.
> 
> Glad it's fixed! The reason why `TIMEOUT` isn't caught is because it's not known whether it's a pass or failure, because we don't know that if we ran the test for long enough it wouldn't eventually pass (i.e. it's a variant of the halting problem). Or from an alternative POV, it's not a failure: it's an error.

If it's uncertain whether or not it has passed, it should absolutely be treated as a failure IMO. But I'm not willing to address it technically so I don't care, lol.

In any case, fixed and ready for review :)

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**
